### DISCUSSION TITLE
Properties Slint (Syntax, Types)

### DISCUSSION THREAD

**Discussion by renlite** [slint-ui/slint#7202] [Q&A]
As far as I know defined properies can be only of primitive types. Is this correct?

Syntax:  For the definition of extra properties you use '\<generic\> type first' definition in opposite to named arguments ('type last').

```Slint
property<int> my-property;
in-out property <bool> checked;

callback select-element(file: string, offset: int, x: length, y: length);
```

Why is the keyword property required? Can a component have other members?

```Slint
// example:
in text: string;
in-out checked: bool;
```

**Comment by mcitem** [comment]
```slint
struct Item {
    id:int,
    name:string,
}
export component Example {
    property <[int]> list;
    property <Item> one-item;
    property <[Item]> items;
    property <{x:int, y:int, z:int}> point;
    property <[{x:int, y:int, z:int}]> points;
    callback select-element(string, int, length, length);
}
```
 defined properties can also be a struct and a list.
 
 If "in-out / in / out" is not added, then this attribute can be used as a private variable. The identifier in the callback is meaningless; it is the same as declaring a callback directly with the type.
 
 And for the type-suffix definition method, if you don't want to append "in-out / in / out", it would conflict with the syntax for setting property values, right?
 ```slint
 //example
text: string;
text: "string";
 ```
 
 Btw, I am not a Slint developer, this is just my personal understanding.

**Reply by renlite** [reply]
Ok, I wasn't precise.  I meant to use component in a property.

Yes, leaving "in-out / in / out" out is not readable but comparing definition of "struct, enum, callback" and "property" looks not consistent.
Defining qualifier as required could make the keyword 'property' obsolet and using type-last definition would feel more consistent?

```Slint
struct Item {
    id: int,
    name: string,
}

export component Example {
  // without keyword
  in-out one-item <Item>: {id: 10, name: "test"};
  out two-item <Item>;
  //with keyword
  out property text <string>;
  in-out property my-item <Item>;
  out property test-item <Item>: {id :10, name: "test"};
}
```
